### DISCUSSION TITLE
Validate that there are no extra fields

### DISCUSSION THREAD

**Discussion by alifemove** [santhosh-tekuri/jsonschema#146] [Q&A]
I'm not entirely sure how jsonschema validates a payload, but if a schema doesn't have a field and the payload does, shouldn't that be an error?

```go
        compiler := jsonschema.NewCompiler()
	compiler.Draft = jsonschema.Draft7
	sch, err := compiler.Compile(cwd + "/metadata/metadata.json")
	if err != nil {
		fmt.Println("ERROR", err.Error())
	}

	if err = sch.Validate(response); err != nil {
		fmt.Println("ERROR", err.Error())
	}
```

If the schema is something like
```json
{
  "type": "object",
  "properties": {
    "Country": {
      "type": "string",
      "default": null
    },
}
```
and the payload has
```json
{
    "Country":"US",
    "State":"OH"
}
```
I would assume it should fail because State is not expected.

For example
In [Fastify](https://fastify.dev/docs/latest/Reference/Validation-and-Serialization/) they have the ability to add a schema onto endpoints and it will validate the payload against the schema and automatically remove fields that shouldn't be in the schema.

If the extra fields at least came back as a list of errors or something I would be able to create a function to loop over the errors removing the fields, but instead I get a valid response.

**Comment by alifemove** [comment]
The problem has been that if the Validate function doesn't recognize what its validating it doesn't error.
For instance, the metadata I'm using is not the entire payload, it's a single record so it needs validated for every db record so if you use it to validate the entire payload nothing happens...

**Comment by santhosh-tekuri** [comment]
@alifemove 

you should use [additionalProperties](https://json-schema.org/understanding-json-schema/reference/object#additionalproperties) keyword to achieve the behaviour you are describing.

for example, if your schema is modified as shown below, I will emit error on unknown properties

```json
{
    "type": "object",
    "properties": {
        "Country": {
          "type": "string",
          "default": null
        },
    },
    "additionalProperties": false,
}
```

**Reply by alifemove** [reply]
Yeah I figured that part out. 
The schemas I'm trying to use are [these](https://github.com/RESOStandards/schema-registry/pull/6/files) one of them uses additionalProperties false and the other is true, so I need to use the restrictive schema.
When you go to validate 
```
		if err = sch.Validate(row); err != nil {
			log.Fatalf("%#v", err)
		}
```
nothing happens when using the permissive schema,  and when you use the restrictive schema it fails for every value passed in to validate. (Trying to validate against the Property definition)
In the schema it has this `"oneOf": [` right near the beginning of the schema and the validations all fail in that section even though I want to validate against the Property definition.

**Reply by alifemove** [reply]
I figured out that I can remove the `oneOf` section and add a `ref: #/definitions/Property` and it will validate the correct section, but I'm not sure how to do it with the `oneOf` being there.


When I do remove `oneOf` and validate I get a bunch of errors like `/NumberOfUnitsTotal/type] expected integer, but got null` even though nullable is true on almost everything.
